### PR TITLE
disposeBag: add a HasDisposeBag protocol

### PR DESCRIPTION
**Pull request by twittemb**
In this commit, we try to conform to protocol oriented programming. It is more flexible to
introduce a DisposeBagable protocol so that every object that respect this one
can have a DisposeBag (NSObject does respect that protocol now)

### CODE REVIEW

File: NSObject_Rx/DisposeBagable.swift
Comment (ashfurrow) [inline NSObject_Rx/DisposeBagable.swift]: Interesting – I didn't expect associated objects to work with non-NSObject subclasses. You've verified this works?

File: Readme.md
Comment (ashfurrow) [inline Readme.md]: Excellent addition to the readme 👍

File: NSObject_RxTests/NSObject_RxTests.swift
Comment (ashfurrow) [inline NSObject_RxTests/NSObject_RxTests.swift]: Good test coverage, awesome!

**Review by ashfurrow — APPROVED** [review]
Looks good to me! Anyone from @RxSwiftCommunity/contributors available to review?

File: NSObject_Rx/DisposeBagable.swift
Comment (twittemb) [reply NSObject_Rx/DisposeBagable.swift]: I will add a Unit test for this

Comment (ashfurrow) [reply NSObject_Rx/DisposeBagable.swift]: Good idea 👍

**Comment by freak4pc** [comment]
My only gotcha with this is I really don't think this is an intuitive protocol name. Definitely an opinionated thing as naming is hard :) 

Great implementation though!

File: NSObject_Rx/DisposeBagable.swift
Comment (twittemb) [reply NSObject_Rx/DisposeBagable.swift]: Done, there is a new Test for that. It works.

**Comment by twittemb** [comment]
For the name, there is another possibility: HasDisposeBag
Is this better ?

**Comment by freak4pc** [comment]
What about `Disposing` (also possible `Disposer` but less common than using `ing` when `able` doesn't make english sense) ? 

It's very tricky since naming conventions are hard in this sense :) 

Opinions ? @RxSwiftCommunity/contributors

File: Readme.md
Comment (freak4pc) [reply Readme.md]: Wondering on why we should call this `rxDisposebag` given the fact it's from RxSwift is rather implicit? what about just calling it disposeBag across the board?

Comment (twittemb) [reply Readme.md]: Good point. I will fix that.

**Comment by freak4pc** [comment]
You should also fix CI :)

**Comment by twittemb** [comment]
I like the HasDisposeBag, because it is very explicit when you read the signature of the class that will implements it.

**Comment by twittemb** [comment]
CI is broken for the Demo app. It was already broken before I think. 
As far I as I can tell, the demo app is just for Unit Testing. I think it could be removed since we can implement Unit Tests directly in the main project. Am I wrong ?

**Comment by freak4pc** [comment]
I'm definitely OK with this @twittemb . 
Master is actually passing green so this seems to be related to the changes you made here

**Comment by ashfurrow** [comment]
Yeah, CI is broken for this PR but it's not your fault! Looks like you're a Carthage user right? We just need to run `pod update NSObject-Rx` inside the Demo directory. I can take care of that if this PR is otherwise good to go. 

EDIT: I've updated the readme to have the CI badge display the latest _master_ build, instead of just the latest build of any branch.

**Comment by freak4pc** [comment]
I think the two pending items are : 
1. Having the associated var named `disposeBag` across the board instead of `rxDisposebag` and updating docs accordingly
2. Gather feedback from community regarding protocol name. Or if there is no feedback for a day or two just leave the current naming which is fine as well :)

**Comment by twittemb** [comment]
Ok. I've made the following changes:
- the var is now named "disposeBag",
- the protocols is now named "HasDisposeBag", it should be a more consensual choice.

**Comment by freak4pc** [comment]
Thanks for the `disposeBag` change. 💯 

Regarding the `Has` prefix, I've never seen a naming convention where that is used for a protocol so I still think it would be best to wait for some feedback from other contributors. And if not we can perhaps just leave `DisposeBagable`. 

`HasDisposeBag` feels a bit off perhaps? I guess i'm not religious about it but just not sure that's proper naming

**Comment by ashfurrow** [comment]
See this is why I liked having all protocols suffixed with "Type", but that went away in Swift 3 :\ I'm in favour of either – we can always change it later + add a `typealias` for the old name.

**Comment by twittemb** [comment]
I was inspired by the following article by Krzysztof Zabłocki, who is involved in the swift community.
http://merowing.info/2017/04/using-protocol-compositon-for-dependency-injection/

He uses the "Has..." to specify every kinds of Services a class can provide (in a DI context)

Another idea: It could be something like "DisposeBagProvider" ... but it is pretty long.

**Comment by freak4pc** [comment]
What about `DisposeBagProvider` (e.g. it provides a dispose bag)

also @ashfurrow RxSwift itself still uses `ObservableConvertibleType` etc so I think that's fine.
Problem is "DisposeBagType" means the object itself is a dispose bag, and not that it is able to dispose thing. 

Interesting about Krzysztof's article. We can keep it this way for now. I'm still a bit iffy about this :) But that's personal taste. 

I'd prefer Disposing (e.g. it can Dispose) or the aforementioned `DisposeBagProvider`. But let's leave it at HasDisposeBag and be done with it I guess !

**Comment by twittemb** [comment]
ok. Thanks for your feedback anyway. 
Just for the record, NSObject-Rx was useful for the project I am doing on my Github. It is about implementing Navigation in iOS apps using a Weaving Pattern. I have experienced many architecture solutions and the best I've found so far was MVVM (with RxSwift) + Coordinator pattern. But I found the Coordinator pattern a bit verbose and there was a lot of boilerplate code to write each time (many delegate patterns). So I combined Coordinator + RxSwift + the idea of navigation state (a bit like Redux) to make a Framework that could ease the implementation of such a pattern. The idea is really to:
- remove navigation code from ViewController for a best separation of concerns,
- promote dependency injection
- allow to cut the storyboards into atomic pieces to ease the reusability (in different navigation contexts) and the collaboration (no more storyboards conflicts).
There is still a lot to do but I have something working. Here is the repo: 
https://github.com/twittemb/Weavy

Feel free to contribute :-)

**Comment by twittemb** [comment]
Hi,
What is the validation process for this PR ? Is there a delay to allow the community to review it ?
thanks.

**Comment by freak4pc** [comment]
I think there's agreement with me and @ashfurrow so that's enough.

Let's just get the CI working and LGTM!

**Comment by ashfurrow** [comment]
Yeah, we should definitely get this merged. Any ideas what's wrong with the tests on Travis? I can help take a look tomorrow, let me know.

**Comment by freak4pc** [comment]
I'm on an earlier time zone, I'll take a look around tomorrow my morning :)

File: Demo/Podfile.lock
Comment (toshi0383) [inline Demo/Podfile.lock]: I think this needs to be `../NSObject_Rx` to make CI happy.

**Comment by freak4pc** [comment]
Hey @twittemb 
Can you remove the Podfile.lock and Cartfile.resolved changes from this commit? I don't think it belongs as part of this change and is probably what causing the CI to fail. 

Once that is changed and merged @ashfurrow can separately release a new version

**Comment by twittemb** [comment]
Hey @freak4pc 
I have made the changes.
IMHO the Cartfile.resolved should be in the repo. This is what allows to do a "carthage bootstrap" instead of a "carthage update" to ensure the consistency of the dependencies versions.

**Comment by freak4pc** [comment]
@twittemb I didn't mean to remove the entire Podfile.lock ... I mean you should not have changed it... Can you fix that ? Right now you just deleted the entire file. Regarding Carthage.resolved, might be but that shouldn't be part of this specific PR IMO

**Comment by twittemb** [comment]
@freak4pc ok sorry for the misunderstanding ... I've restored the Podfile.lock with the changes you recommend.

**Comment by freak4pc** [comment]
This PR still introduces Podfile.lock changes... 
My point is this PR shouldn't have any changes on Podfile.lock or Cartfile.resolved at the moment. 

e392605a-3c39-4066-a038-abe6b7ef54bb

**Comment by twittemb** [comment]
@freak4pc should be alright now :-)

**Comment by freak4pc** [comment]
@twittemb Yup! looks good! Let's see if CI clears and I'm good to merge this. 
Thanks for your patience and cooperation !

File: Demo/Podfile.lock
Comment (freak4pc) [reply Demo/Podfile.lock]: No, this points to the podspec so it should be fine

**Comment by freak4pc** [comment]
Very confusing 🤔 
Master is passing and there's only code changes, so not sure why the module isn't passing. 

@ashfurrow any thoughts here? I didn't setup the CI here so not sure if any assumptions I'm missing

**Comment by ashfurrow** [comment]
It's possible that Travis has changed something on their end. I'll investigate and get back to everyone.

**Comment by twittemb** [comment]
ok thanks @ashfurrow 
If this can help, I can still make a first PR that would remove Demo app and integrate the tests that were done in it directly into the main NSObject-Rx project. I could rebase the current PR on this new one.

**Comment by freak4pc** [comment]
@twittemb If you have the time that would be great! Might also resolve our issue here altogether.

**Comment by twittemb** [comment]
@freak4pc @ashfurrow I'll try to do it today (Montreal time zone)

**Comment by ashfurrow** [comment]
I can reproduce the problem locally – will have a fix shortly. We might as well keep this PR then.

**Comment by ashfurrow** [comment]
Okay, so this looks like an instance of a pretty common issue – Carthage and CocoaPods not being in sync. Happens all the time on Moya :) I've opened #42 with the changes from this PR as well as the changes needed to sort things out.

**Comment by ashfurrow** [comment]
Thanks a lot for contributing! I've invited you to join the RxSwiftCommunity GitHub organization – no pressure to accept! If you'd like more information on what that means, check out our [contributor guidelines](https://github.com/RxSwiftCommunity/contributors) and feel free to reach out with any questions.

**Pull request merged**
